aptos-move/framework/move-stdlib/sources/configs/features.move

```diff
@@ -260,30 +260,67 @@ module std::features {
     public fun sponsored_automatic_account_creation_enabled(): bool acquires Features {
         is_enabled(SPONSORED_AUTOMATIC_ACCOUNT_CREATION)
     }
+    /// Whether reconfiguration with DKG is enabled.
+    /// Lifetime: transient
+    const RECONFIGURE_WITH_DKG: u64 = 36;
+
+    public fun get_reconfigure_with_dkg_feature(): u64 { RECONFIGURE_WITH_DKG }
+
+    public fun reconfigure_with_dkg_enabled(): bool acquires Features {
+        is_enabled(RECONFIGURE_WITH_DKG)
+    }
+
     // ============================================================================================
     // Feature Flag Implementation
 
     /// The provided signer has not a framework address.
     const EFRAMEWORK_SIGNER_NEEDED: u64 = 1;
 
     /// The enabled features, represented by a bitset stored on chain.
-    struct Features has key {
+    struct Features has copy, drop, key, store {
         features: vector<u8>,
     }
 
     /// Function to enable and disable features. Can only be called by a signer of @std.
     public fun change_feature_flags(framework: &signer, enable: vector<u64>, disable: vector<u64>)
     acquires Features {
         assert!(signer::address_of(framework) == @std, error::permission_denied(EFRAMEWORK_SIGNER_NEEDED));
-        if (!exists<Features>(@std)) {
-            move_to<Features>(framework, Features { features: vector[] })
+        if (reconfigure_with_dkg_enabled()) {
+            let features = if (config_for_next_epoch::does_exist<Features>()) {
+                config_for_next_epoch::copied<Features>()
+            } else if (exists<Features>(@std)) {
+                *borrow_global<Features>(@std)
+            } else {
+                Features { features: vector[] }
+            };
+            apply_diff(&mut features, enable, disable);
+            config_for_next_epoch::upsert(framework, features);
+        } else {
+            if (!exists<Features>(@std)) {
+                move_to<Features>(framework, Features { features: vector[] })
+            };
+            let features = borrow_global_mut<Features>(@std);
+            apply_diff(features, enable, disable);
         };
-        let features = &mut borrow_global_mut<Features>(@std).features;
+    }
+
+    /// Apply all the pending feature flag changes. Should only be used at the end of a reconfiguration with DKG.
+    ///
+    /// While the scope is public, it can only be usd in system transactions like `block_prologue` and governance proposals,
+    /// who have permission to set the flag that's checked in `extract()`.
+    public fun on_new_epoch() acquires Features {
+        if (config_for_next_epoch::does_exist<Features>()) {
+            let features = config_for_next_epoch::extract<Features>();
+            *borrow_global_mut<Features>(@std) = features;
+        }
+    }
```

Comment:
`std::features::Features` is a critical on-chain state that i think has to have a pending change buffer (and therefore a `on_new_epoch` function).
What's special is, it's defined in `move-stdlib` package (while `reconfiguration.move` and all the other configs I identified so far are in `move-framework` package). for all other configs' `on_new_epoch()` , we can protect it by setting scope to `public(friend)` and make `reconfiguration.move` a friend.
But for Features, this protection is not available.

The workaround I'm thinking.
- Leave this function `public`.
- Introduce a flag struct `ExtractPermit`.
- Any system txn that may invoke `reconfigure` should put `0x1::ExtractPermit` before the invocation, and remove it after.
- Inside `config_for_next_epoch::extract`, do the work only if the `0x1::ExtractPermit` exists.

Reply:
Potential alternative:
- Have a clone of `reconfigure()` that takes a signer: `reconfigure_v2(account: &signer)`.
- Update all reconfigure callers to use `reconfigure_v2`.
- Have `features::on_new_epoch()` requires a signer and assert the signer is 0x0/0x1.